File: src/pymax/dispatch/router.py
from __future__ import annotations

from collections import defaultdict
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from enum import Enum
from typing import TYPE_CHECKING, Any, Generic, TypeAlias, TypeVar

from pymax.types import MessageDeleteEvent

from .enums import EventType

if TYPE_CHECKING:
    from pymax import Client
    from pymax.base import BaseClient
    from pymax.protocol import InboundFrame
    from pymax.types import Chat
    from pymax.types.domain import Message
    from pymax.types.events import (
        MessageReadEvent,
        PresenceEvent,
        ReactionUpdateEvent,
        TypingEvent,
    )


class ErrorScope(str, Enum):
    GLOBAL = "global"
    LOCAL = "local"


_EventT = TypeVar("_EventT")
ClientT = TypeVar("ClientT", bound="BaseClient")

HandlerCallback: TypeAlias = Callable[
    [_EventT, ClientT],
    Awaitable[Any] | Any,
]
HandlerDecorator: TypeAlias = Callable[
    [HandlerCallback[_EventT, ClientT]],
    HandlerCallback[_EventT, ClientT],
]
FilterCallback: TypeAlias = Callable[
    [_EventT],
    Awaitable[bool] | bool,
]
StartCallback: TypeAlias = Callable[
    [ClientT],
    Awaitable[Any] | Any,
]
StartDecorator: TypeAlias = Callable[
    [StartCallback[ClientT]],
    StartCallback[ClientT],
]


@dataclass(slots=True)
class ErrorContext(Generic[ClientT]):
    client: ClientT
    event_type: EventType
    event: Any
    handler: HandlerEntry[Any, ClientT] | StartCallback | None
    router: Router[ClientT]


ErrorCallback: TypeAlias = Callable[
    [Exception, ErrorContext[ClientT]],
    Awaitable[Any] | Any,
]

ErrorDecorator: TypeAlias = Callable[
    [ErrorCallback[ClientT]],
    ErrorCallback[ClientT],
]

DisconnectCallback: TypeAlias = Callable[
    [Exception, bool, float],
    Awaitable[Any] | Any,
]

DisconnectDecorator: TypeAlias = Callable[
    [DisconnectCallback],
    DisconnectCallback,
]


@dataclass(slots=True)
class HandlerEntry(Generic[_EventT, ClientT]):
    callback: HandlerCallback[_EventT, ClientT]
    filters: tuple[FilterCallback[_EventT], ...] = ()


@dataclass(slots=True)
class ErrorEntry(Generic[ClientT]):
    callback: ErrorCallback[ClientT]
    scope: ErrorScope = ErrorScope.GLOBAL


ErrorSource: TypeAlias = HandlerEntry[Any, ClientT] | StartCallback[ClientT]


class Router(Generic[ClientT]):
    """Контейнер обработчиков событий PyMax.

    Роутер хранит обработчики и фильтры. Когда приходит событие, dispatcher
    проходит по root router и его дочерним роутерам, проверяет фильтры и
    вызывает подходящие callbacks как ``handler(event, client)``.

    Example:
        .. code-block:: python

           from pymax import Client, Message, Router

           router = Router[Client]()

           def is_start(message: Message) -> bool:
               return message.text == "/start"

           @router.on_message(is_start)
           async def start(message: Message, client: Client) -> None:
               await message.answer("Привет")

           client = Client(phone="+79990000000")
           client.include_router(router)
    """

    def __init__(self) -> None:
        self.handlers: dict[
            EventType,
            list[HandlerEntry[Any, ClientT]],
        ] = defaultdict(list)

        self.children: list[Router[ClientT]] = []
        self.on_start_handlers: list[StartCallback[ClientT]] = []
        self.error_handlers: list[ErrorEntry[ClientT]] = []
        self.disconnect_handlers: list[DisconnectCallback] = []

    def on_error(
        self,
        scope: ErrorScope = ErrorScope.GLOBAL,
    ) -> ErrorDecorator[ClientT]:
        def decorator(callback: ErrorCallback[ClientT]) -> ErrorCallback[ClientT]:
            self.error_handlers.append(ErrorEntry(callback=callback, scope=scope))
            return callback

        return decorator

    def on_disconnect(self) -> DisconnectDecorator:
        """Регистрирует обработчик потери соединения.

        Callback вызывается как ``handler(exception, reconnect, delay)``:
        исходная ошибка, будет ли reconnect и задержка перед ним.
        """

        def decorator(callback: DisconnectCallback) -> DisconnectCallback:
            self.disconnect_handlers.append(callback)
            return callback

        return decorator

    def on(
        self,
        event: EventType,
        /,
        *filters: FilterCallback[_EventT],
    ) -> HandlerDecorator[_EventT, ClientT]:
        """Регистрирует обработчик события по ``EventType``.

        Args:
            event: Тип события.
            *filters: Фильтры, которые получают событие и возвращают bool.

        Returns:
            Декоратор для ``handler(event, client)``.

        Example:
            .. code-block:: python

               from pymax import Client
               from pymax.protocol import InboundFrame

               @router.on(EventType.RAW)
               async def raw(frame: InboundFrame, client: Client) -> None:
                   print(frame.payload)
        """

        def decorator(
            handler: HandlerCallback[_EventT, ClientT],
        ) -> HandlerCallback[_EventT, ClientT]:
            self.handlers[event].append(
                HandlerEntry(
                    callback=handler,
                    filters=filters,
                )
            )
            return handler

        return decorator

    def include_router(self, router: Router[ClientT]) -> None:
        """Подключает дочерний роутер.

        Args:
            router: Router, обработчики которого нужно добавить в дерево.

        Returns:
            ``None``.
        """
        self.children.append(router)

    def on_start(self) -> StartDecorator:
        """Регистрирует обработчик старта клиента.

        Returns:
            Декоратор для ``handler(client)``.
        """

        def decorator(handler: StartCallback) -> StartCallback:
            self.on_start_handlers.append(handler)
            return handler

        return decorator

    def on_message(
        self,
        *filters: FilterCallback[Message],
    ) -> HandlerDecorator[Message, ClientT]:
        """Регистрирует обработчик новых сообщений.

        Args:
            *filters: Фильтры для ``Message``.

        Returns:
            Декоратор для ``handler(message, client)``.
        """
        return self.on(EventType.MESSAGE_NEW, *filters)

    def on_message_edit(
        self,
        *filters: FilterCallback[Message],
    ) -> HandlerDecorator[Message, ClientT]:
        """Регистрирует обработчик редактирования сообщений.

        Args:
            *filters: Фильтры для ``Message``.

        Returns:
            Декоратор для ``handler(message, client)``.
        """
        return self.on(EventType.MESSAGE_EDIT, *filters)

    def on_message_delete(
        self,
        *filters: FilterCallback[MessageDeleteEvent],
    ) -> HandlerDecorator[MessageDeleteEvent, ClientT]:
        """Регистрирует обработчик удаления сообщений.

        Args:
            *filters: Фильтры для ``MessageDeleteEvent``.

        Returns:
            Декоратор для ``handler(event, client)``.
        """
        return self.on(EventType.MESSAGE_DELETE, *filters)

    def on_message_read(
        self,
        *filters: FilterCallback[MessageReadEvent],
    ) -> HandlerDecorator[MessageReadEvent, ClientT]:
        """Регистрирует обработчик изменения отметки прочтения."""
        return self.on(EventType.MESSAGE_READ, *filters)

    def on_typing(
        self,
        *filters: FilterCallback[TypingEvent],
    ) -> HandlerDecorator[TypingEvent, ClientT]:
        """Регистрирует обработчик набора текста."""
        return self.on(EventType.TYPING, *filters)

    def on_presence(
        self,
        *filters: FilterCallback[PresenceEvent],
    ) -> HandlerDecorator[PresenceEvent, ClientT]:
        """Регистрирует обработчик изменения присутствия пользователя."""
        return self.on(EventType.PRESENCE, *filters)

    def on_reaction_update(
        self,
        *filters: FilterCallback[ReactionUpdateEvent],
    ) -> HandlerDecorator[ReactionUpdateEvent, ClientT]:
        """Регистрирует обработчик обновления реакций сообщения."""
        return self.on(EventType.REACTION_UPDATE, *filters)

    def on_chat_update(
        self,
        *filters: FilterCallback[Chat],
    ) -> HandlerDecorator[Chat, ClientT]:
        """Регистрирует обработчик обновления чата.

        Args:
            *filters: Фильтры для ``Chat``.

        Returns:
            Декоратор для ``handler(chat, client)``.
        """
        return self.on(EventType.CHAT_UPDATE, *filters)

    def on_raw(
        self,
        *filters: FilterCallback[InboundFrame],
    ) -> HandlerDecorator[InboundFrame, ClientT]:
        """Регистрирует обработчик исходных frame-ов.

        Args:
            *filters: Фильтры для ``InboundFrame``.

        Returns:
            Декоратор для ``handler(frame, client)``.
        """
        return self.on(EventType.RAW, *filters)


ClientRouter: TypeAlias = Router["Client"]

File: src/PHPMax/Dispatch/Router.php
<?php

declare(strict_types=1);

namespace PHPMax\Dispatch;

use PHPMax\Api\Binding;
use PHPMax\Protocol\InboundFrame;
use PHPMax\Runtime\App;
use Throwable;

class Router
{
    /** @var array<string, list<array{0: callable, 1: list<callable>}>> */
    private $handlers = [];
    /** @var list<callable> */
    private $startHandlers = [];
    /** @var list<Router> */
    private $children = [];
    /** @var list<array{0: callable, 1: string}> */
    private $errorHandlers = [];
    /** @var list<callable> */
    private $disconnectHandlers = [];
    /** @var EventResolver */
    private $resolver;
    /** @var EventMapper */
    private $mapper;

    public function __construct(?EventResolver $resolver = null, ?EventMapper $mapper = null)
    {
        $this->resolver = $resolver ?: new EventResolver();
        $this->mapper = $mapper ?: new EventMapper();
    }

    public function on(string $eventType, callable $handler, callable ...$filters): self
    {
        if (!isset($this->handlers[$eventType])) {
            $this->handlers[$eventType] = [];
        }
        $this->handlers[$eventType][] = [$handler, $filters];

        return $this;
    }

    public function onRaw(callable $handler, callable ...$filters): self
    {
        return $this->on(EventType::RAW, $handler, ...$filters);
    }

    public function onMessage(callable $handler, callable ...$filters): self
    {
        return $this->on(EventType::MESSAGE_NEW, $handler, ...$filters);
    }

    public function onMessageEdit(callable $handler, callable ...$filters): self
    {
        return $this->on(EventType::MESSAGE_EDIT, $handler, ...$filters);
    }

    public function onMessageDelete(callable $handler, callable ...$filters): self
    {
        return $this->on(EventType::MESSAGE_DELETE, $handler, ...$filters);
    }

    public function onMessageRead(callable $handler, callable ...$filters): self
    {
        return $this->on(EventType::MESSAGE_READ, $handler, ...$filters);
    }

    public function onTyping(callable $handler, callable ...$filters): self
    {
        return $this->on(EventType::TYPING, $handler, ...$filters);
    }

    public function onPresence(callable $handler, callable ...$filters): self
    {
        return $this->on(EventType::PRESENCE, $handler, ...$filters);
    }

    public function onReactionUpdate(callable $handler, callable ...$filters): self
    {
        return $this->on(EventType::REACTION_UPDATE, $handler, ...$filters);
    }

    public function onChatUpdate(callable $handler, callable ...$filters): self
    {
        return $this->on(EventType::CHAT_UPDATE, $handler, ...$filters);
    }

    public function onFileReady(callable $handler, callable ...$filters): self
    {
        return $this->on(EventType::FILE_READY, $handler, ...$filters);
    }

    public function onVideoReady(callable $handler, callable ...$filters): self
    {
        return $this->on(EventType::VIDEO_READY, $handler, ...$filters);
    }

    public function onError(callable $handler, string $scope = ErrorScope::GLOBAL): self
    {
        if ($scope !== ErrorScope::GLOBAL && $scope !== ErrorScope::LOCAL) {
            throw new \InvalidArgumentException('Invalid error scope: ' . $scope);
        }
        $this->errorHandlers[] = [$handler, $scope];

        return $this;
    }

    public function onDisconnect(callable $handler): self
    {
        $this->disconnectHandlers[] = $handler;

        return $this;
    }

    public function onStart(callable $handler): self
    {
        $this->startHandlers[] = $handler;

        return $this;
    }

    public function includeRouter(Router $router): self
    {
        $this->children[] = $router;

        return $this;
    }

    /**
     * @param mixed $client
     */
    public function emitStart($client): void
    {
        $this->emitStartInTree($client, $this);
    }

    /**
     * @param mixed $client
     */
    private function emitStartInTree($client, Router $root): void
    {
        foreach ($this->startHandlers as $handler) {
            try {
                $handler($client);
            } catch (Throwable $e) {
                $handled = $root->emitError($e, EventType::ON_START, null, $client, $this, $handler);
                if (!$handled) {
                    throw $e;
                }
            }
        }
        foreach ($this->children as $child) {
            $child->emitStartInTree($client, $root);
        }
    }

    /**
     * @param mixed $event
     * @param mixed $client
     */
    public function dispatch(string $eventType, $event, $client): void
    {
        $this->dispatchInTree($eventType, $event, $client, $this);
    }

    /**
     * @param mixed $event
     * @param mixed $client
     */
    private function dispatchInTree(string $eventType, $event, $client, Router $root): void
    {
        foreach ($this->handlers[$eventType] ?? [] as $entry) {
            [$handler, $filters] = $entry;
            try {
                if (!$this->matches($filters, $event)) {
                    continue;
                }
                $handler($event, $client);
            } catch (Throwable $e) {
                $handled = $root->emitError($e, $eventType, $event, $client, $this, $handler);
                if (!$handled) {
                    throw $e;
                }
            }
        }

        foreach ($this->children as $child) {
            $child->dispatchInTree($eventType, $event, $client, $root);
        }
    }

    /**
     * @param mixed $client
     */
    public function dispatchRaw(InboundFrame $frame, $client): void
    {
        $this->dispatch(EventType::RAW, $frame, $client);
    }

    /**
     * @param mixed $client
     */
    public function dispatchFrame(InboundFrame $frame, $client, bool $dispatchRaw = true): void
    {
        $eventType = $this->resolver->resolve($frame);
        if ($eventType !== null) {
            $event = $this->mapper->map($eventType, $frame);
            $this->dispatch($eventType, $this->bindEvent($event, $client), $client);
        }

        if ($dispatchRaw) {
            $this->dispatchRaw($frame, $client);
        }
    }

    /**
     * @param mixed $event
     * @param mixed $client
     */
    public function emitError(Throwable $exception, string $eventType, $event, $client, Router $failedRouter, ?callable $handler): bool
    {
        $handled = false;
        $context = new ErrorContext($client, $eventType, $event, $failedRouter, $handler);
        foreach ($this->errorEntries() as $entry) {
            /** @var Router $owner */
            $owner = $entry[0];
            /** @var callable $errorHandler */
            $errorHandler = $entry[1];
            /** @var string $scope */
            $scope = $entry[2];

            if ($scope === ErrorScope::LOCAL && $owner !== $failedRouter) {
                continue;
            }

            $handled = true;
            try {
                $errorHandler($exception, $context);
            } catch (Throwable $e) {
                return false;
            }
        }

        return $handled;
    }

    public function emitDisconnect(Throwable $exception, bool $reconnect = false, float $delay = 0.0): void
    {
        foreach ($this->disconnectHandlers as $handler) {
            try {
                $handler($exception, $reconnect, $delay);
            } catch (Throwable $e) {
                // Disconnect handlers must not interrupt reconnect/close flow.
            }
        }

        foreach ($this->children as $child) {
            $child->emitDisconnect($exception, $reconnect, $delay);
        }
    }

    /**
     * @param list<callable> $filters
     * @param mixed $event
     */
    private function matches(array $filters, $event): bool
    {
        foreach ($filters as $filter) {
            if (!$filter($event)) {
                return false;
            }
        }

        return true;
    }

    /**
     * @param mixed $event
     * @param mixed $client
     * @return mixed
     */
    private function bindEvent($event, $client)
    {
        if (!is_object($client) || !method_exists($client, 'app')) {
            return $event;
        }

        $app = $client->app();
        if (!$app instanceof App) {
            return $event;
        }

        return Binding::bindApiModel($app, $event);
    }

    /**
     * @return list<array{0: Router, 1: callable, 2: string}>
     */
    private function errorEntries(): array
    {
        $entries = [];
        foreach ($this->errorHandlers as $entry) {
            $entries[] = [$this, $entry[0], $entry[1]];
        }
        foreach ($this->children as $child) {
            foreach ($child->errorEntries() as $entry) {
                $entries[] = $entry;
            }
        }

        return $entries;
    }
}
